feat: add live staking data with geo-blocking support

amitabh94 opened the pull request:
## **Description**

This PR replaces mock staking data with actual staking positions for users and integrates the Staking SDK. It also implements geo-blocking based on user location and displays real vault data.

Changes made:

- Integrated the Staking SDK to fetch and display actual user staking positions
- Implemented geo-blocking to restrict access based on region
- Updated the UI to display live vault data

Additional Context:

- Replaced mock data that was previously used in staking-related views
- Users will now see their actual staking positions within the app
- Geo-blocking ensures compliance with regional restrictions
- Vault data is dynamically fetched and displayed in real-time

## **Related issues**

Fixes: [STAKE 785](https://consensyssoftware.atlassian.net/jira/software/projects/STAKE/boards/550/backlog?selectedIssue=STAKE-785), [STAKE 806](https://consensyssoftware.atlassian.net/browse/STAKE-806)

## **Manual testing steps**

1. Enable the staking feature flag in js.env locally.
2. Log in as a user with staking positions and verify that the staking section shows live data.
3. Access the staking feature from a geo-blocked region to verify the geo-blocking functionality.
4. Ensure that vault data is correctly displayed based on the user’s chain and staking status.

## **Screenshots/Recordings**


https://github.com/user-attachments/assets/4bf1589c-7cda-441f-ab44-05ada1ce80aa



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

amitabh94 committed:
refactor: move StakingEarnings component to the stake directory

amitabh94 committed:
chore: update @metamask/stake-sdk to version 0.2.13

amitabh94 committed:
feat: update stakeSdkProvider to use new stakingApiService

The stakeSdkProvider component has been updated to use the new stakingApiService from the @metamask/stake-sdk library.

amitabh94 committed:
refactor: update useStakeContext to throw error when used outside StakeProvider

amitabh94 committed:
feat: add new hooks for staking functionality

- Added usePooledStakes hook to fetch pooled stakes data from the staking API service
- Added useStakingChain hook to check if the current chain is supported for staking
- Added useStakingEligibility hook to fetch staking eligibility data from the staking API service
- Added useVaultData hook to fetch vault data from the staking API service

These hooks provide the necessary functionality for staking operations and data retrieval in the UI.

amitabh94 committed:
feat: update StakingBalance component to replace mock data

amitabh94 committed:
feat: update StakingEarnings component to use new hooks and provider

amitabh94 committed:
feat: remove unused useGetStakingEligibility hook and stakingApi slice

amitabh94 committed:
refactor: update StakeInputView to use new hooks and remove unused code

amitabh94 committed:
refactor: update import path for mock data in StakeConfirmationView

amitabh94 committed:
refactor: update UnstakeInputView.test.tsx to use new hooks and mock data

amitabh94 committed:
feat: update StakeButton to use new hooks and provider

amitabh94 committed:
feat: update TokenListItem to use new hooks for staking

amitabh94 committed:
feat: update usePooledStakes hook to include additional status checks

amitabh94 committed:
refactor: rename useIsStakingSupportedChain to useStakingChain

amitabh94 committed:
refactor: remove unused code in useStakingEligibility hook

amitabh94 committed:
refactor: optimize useVaultData hook and add annual reward rate calculation

amitabh94 committed:
refactor: optimize useBalance hook and add staked balance calculations

amitabh94 committed:
feat: create useStakingEarnings hook for calculating staking earnings

This commit adds a new hook called useStakingEarnings to calculate staking earnings. The hook utilizes the useVaultData, useBalance, and usePooledStakes hooks to fetch necessary data and perform calculations. It calculates the lifetime rewards in ETH and fiat, as well as the estimated annual earnings in ETH and fiat.

amitabh94 committed:
refactor: optimize useStakingInput hook and improve annual rewards calculation

amitabh94 committed:
feat: replace staking earnings mock data with api data

amitabh94 committed:
feat: add staking earnings mock data

amitabh94 committed:
refactor: optimize EstimatedAnnualRewardsCard component and add loading skeleton

amitabh94 committed:
feat: replace staking mocked asset data with actual values

amitabh94 committed:
refactor: remove unused state and optimize TokenDetails component

amitabh94 committed:
refactor: optimize StakingButtons component and update props

amitabh94 committed:
refactor: update reward rate display in RewardsCard component

amitabh94 committed:
feat: update UnstakeInputView component and use actual staked balance

amitabh94 committed:
refactor: update StakeInputView component and optimize estimated rewards display

amitabh94 committed:
refactor: update TokenListItem component to use updated staking chain hook

amitabh94 committed:
refactor: update StakeConfirmationView to use actual staked balance and reward data

amitabh94 committed:
refactor: move stake button component to stake directory

amitabh94 committed:
refactor: update import path for Theme type in StakingEarnings.styles.tsx

amitabh94 commented:
@SocketSecurity ignore-all

@metamask/Stake-sdk uses fetch to communicate with staking backend api service

siibars commented on app/components/UI/Stake/sdk/stakeSdkProvider.tsx:
```diff
@@ -11,30 +16,35 @@ import React, {
   useEffect(() => {
     (async () => {
       try {
+        setStakingApiService(SDK.stakingApiService);
```
hi, since the stake-sdk staking api service is currently only handling pooled staking you could move this line in the if (sdkType === StakingType?.POOLED) condition.

amitabh94 replied:
Good point. I also thought about this and intentionally kept it outside the pooled block as the base url is going to be the same for different staking services and maybe we will add other methods to the same service instead of creating different api services for different staking types.

That said, if in the future we plan to have different api services then we can refactor this but my assumption is we will keep the same.

siibars commented on app/components/UI/Stake/components/StakeButton/index.tsx:
```diff
@@ -93,3 +93,11 @@ export const StakeButton = ({ asset }: StakeButtonProps) => {
 };
+
+export const StakeButton = (props: StakeButtonProps) => (
+  <StakeSDKProvider>
```
hi, you may want to move up in the component hierarchy tree so that we have as less as possible of StakeSdkProvider instance, for ex [here](https://github.com/MetaMask/metamask-mobile/blob/e68b2a59355533f985b0f47786011a92bccc10a4/app/components/Nav/Main/MainNavigator.js#L118) or [here](https://github.com/MetaMask/metamask-mobile/blob/07837648c4b4b5ab08927204caad6cd7f8062b06/app/components/Views/Wallet/index.tsx#L604)

amitabh94 replied:
I thought of moving them up the component hierarchy but was a bit reluctant to add them for components not owned by the staking team. IIRC, these are the only two instances where I have wrapped them in the provider, but in the future if we feel that we need to expand context then we can move them up in the hierarchy.

siibars commented on app/components/UI/Stake/hooks/usePooledStakes.ts:
```diff
@@ -0,0 +1,120 @@
+        setPooledStakesData(accounts[0] || null);
+        setExchangeRate(fetchedExchangeRate);
+      } catch (err) {
+        setError('Failed to fetch pooled stakes');
```
hey, also you may want to use and define the StakeSDKProvider.context.sdkError if you want to have different component handling this error.

amitabh94 replied:
I was of the view that we were using that state mostly for errors that occur during sdk initialization and hence tried to keep this separate from api errors.

amitabh94 committed:
refactor: mock network configuration in StakingBalance and StakingEarnings tests

amitabh94 committed:
feat: rename stake button label to use "Earn" instead of "Stake"

amitabh94 committed:
refactor: update FooterButtonGroup component in StakingConfirmation

The FooterButtonGroup component in StakingConfirmation has been updated to include the use of the `usePooledStakes` hook for refreshing pooled stakes. Additionally, the `navigate` function has been replaced with the `navigation.goBack()` function for the cancel button onPress event.

amitabh94 committed:
refactor: mock usePooledStakes hook in stake tests

amitabh94 committed:
refactor: mock network configuration in Tokens test

amitabh94 committed:
feat: update test snapshot to use "Earn" instead of "Stake"

amitabh94 committed:
refactor: update button label to use "Continue" instead of "Confirm"

amitabh94 committed:
feat: update StakeButton to refresh pooled staking eligibility on button press

- This change ensures that the staking eligibility is always up to date when the user interacts with the stake button.

amitabh94 committed:
chore: revert changes to yarn.lock

amitabh94 committed:
feat: add test for useBalance hook

amitabh94 committed:
feat: add useStakingChain hook test

amitabh94 committed:
feat: add usePooledStakes hook test

amitabh94 committed:
feat: add useVaultData hook test and update annual reward rate calculation

amitabh94 committed:
feat: add useStakingEligibility hook test

amitabh94 committed:
refactor: update variable name for annual rewards in StakeInputView and useStakingInputHooks

amitabh94 committed:
feat: add useStakingEarnings hook test

amitabh94 commented on app/components/UI/AssetOverview/StakingEarnings/StakingEarnings.test.tsx:
This component has been moved to the stake directory.

amitabh94 committed:
Merge branch 'main' into STAKE-806--fe-integrate-stake-and-vault-data

tommasini commented on app/components/UI/Stake/components/StakingConfirmation/ConfirmationFooter/FooterButtonGroup/FooterButtonGroup.test.tsx:
```diff
@@ -89,8 +96,7 @@ describe('FooterButtonGroup', () => {
 
-    expect(mockNavigate).toHaveBeenCalledTimes(1);
-    expect(mockNavigate).toHaveBeenCalledWith('Asset');
+    expect(mockGoBack).toHaveBeenCalledTimes(2);
```
Why go back is happening twice?

amitabh94 replied:
We have had an active discussion as to whether we want to cancel the stake flow or just go back to the input screen. Right now this takes the user back to main screen skipping the input screen.
We just agreed that we will take the user back to the input screen so was planning to change this in the upcoming PR but can do it now as well.

tommasini commented on app/components/UI/Stake/components/StakingEarnings/index.tsx:
```diff
@@ -3,49 +3,60 @@ import { View } from 'react-native';
+
+  const { isStakingSupportedChain } = useStakingChain();
+
+  const isLoadingData = isLoadingEligibility || isLoadingEarningsData;
```
By the name of the variable I believe you mean && instead of || right? The data is only loading if it's eligble? I can be missing some context, but I think it's safe to assume that if is not eligible is never loading if it's eligible can be loading? 
Let me know if I'm missing some context here
```suggestion
  const isLoadingData = isLoadingEligibility && isLoadingEarningsData;
```

amitabh94 replied:
Good catch !

amitabh94 replied:
The isLoadingData is checking whether any of these eligibility functions and data functions results are still being fetched . 
If any of them are still loading then we don't show the data to the user .

tommasini replied:
Thanks for the context!

amitabh94 committed:
feat: redirect user back to input screen on cancel button press

tommasini commented on app/components/UI/Tokens/TokenList/TokenListItem/index.tsx:
```diff
@@ -211,7 +214,9 @@ export const TokenListItem = ({
           </Text>
           {/** Add button link to Portfolio Stake if token is mainnet ETH */}
-          {asset.isETH && isMainnet && <StakeButton asset={asset} />}
+          {asset.isETH && isStakingSupportedChain && (
```
Should tis be under feature flag `isPooledStakingFeatureEnabled()`?

amitabh94 replied:
This is currently already present in the app which renders the stake button you see alongside the ethereum token. 
Within the component , we have enabled the feature flag to enable the user to access this new flow instead of getting directed to the mobile browser for staking.

amitabh94 commented on app/components/UI/Stake/components/StakingConfirmation/ConfirmationFooter/FooterButtonGroup/FooterButtonGroup.tsx:
```diff
@@ -78,12 +79,15 @@ const FooterButtonGroup = ({ valueWei, action }: FooterButtonGroupProps) => {
-        onPress={() => navigate('Asset')}
+        onPress={() => {
+          navigation.goBack();
+          navigation.goBack();
```
I have changed this to one instance based on my comment above.

amitabh94 committed:
Merge branch 'main' into STAKE-806--fe-integrate-stake-and-vault-data

amitabh94 committed:
Merge branch 'main' into STAKE-806--fe-integrate-stake-and-vault-data

amitabh94 committed:
Merge branch 'main' into STAKE-806--fe-integrate-stake-and-vault-data
